google-cloud-storage/lib/google/cloud/storage/bucket/encryption.rb

```diff
@@ -0,0 +1,138 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "google/apis/storage_v1"
+
+module Google
+  module Cloud
+    module Storage
+      class Bucket
+        ##
+        # # Encryption Configuration
+        #
+        # A builder for Google Cloud Storage encryption configurations, passed
+        # to {Bucket#encryption=} in block arguments to {Project#create_bucket}
+        # and {Bucket#update}. See {Project#encryption} for creating instances.
+        #
+        # @see https://cloud.google.com/kms/docs/ Cloud Key Management Service
+        #   Documentation
+        #
+        # @example Encrypt a new file with a default KMS key:
+        #   require "google/cloud/storage"
+        #
+        #   storage = Google::Cloud::Storage.new
+        #
+        #   kms_key_name = "projects/a/locations/b/keyRings/c/cryptoKeys/d"
+        #   encryption = storage.encryption default_kms_key: kms_key_name
```

Comment:
This pattern comes from BigQuery, and I believe the reasoning is that more fields may be added to `Encryption`. However, I like your suggestion. @blowmage? @tswast?

Reply:
Yes. That's the reason. At least for BigQuery they want to keep open the possibility of (and are actively working on AFAIK) other encryption methods besides KMS.

Reply:
Thanks both, I'm checking in with gcs team for more information. 

Follow-up:
Depending on response from the gcs team, potentially, the client library could use a Hash to define this object instead of a helper class.

Reply:
Will raise again in KMS sync on Monday. Please don't merge or release today or this weekend.

Reply:
Will still raise the original question today, but had a chat with @quartzmo about the design.

Given the way the storage client is for Ruby, I'd expect to see:
`bucket.default_kms_key_name = "kms/key" 

Instead of:
`encryption = storage.encryption default_kms_key: kms_key_name`

The secondary feels incorrect, I don't think the GCS client should mirror the BigQuery client on this design. It may work for the BigQuery client but not so much for the GCS client.

Reply:
If this class does stay, I would fully expect to see CSEK options in this class, which I'm not seeing currently.

Reply:
Thanks for raising that concern @tswast. It's a good point, I think we should remove this class all together. I think it over complicated the implementation for CMEK.

Then, the client can be simplified to:
`bucket.default_kms_key_name = "kms/key"`

Reply:
@frankyn Will do.

Reply:
Response for possible future of the encryption object in the Object Resource:
There's discussion for adding additional encryption types and that's why it's defined as an object. Design TBD at this moment. I don't think we should move away from the current decision.

Reply:
Updated. Names are `Bucket#kms_key_name` and `#kms_key_name=`.

Reply:
clarifying previous comment: `Bucket#default_kms_key` and `Bucket#default_kms_key=` were used instead.